[Alvin Tay Ming Hwee] iP

atmh opened the pull request:
# Echo
> "The mind is not a vessel to be filled but a fire to be kindled." – Plutarch ([source](https://medium.com/thethursdaythought/the-mind-is-not-a-vessel-to-be-filled-but-a-fire-to-be-kindled-8a3bedb7c4c4))

Echo frees your mind of having to remember tasks. It's,
- text-based
- easy to learn
- ~~FAST~~ _BLAZINGLY_ FAST to use

All you need to do is,
1. download it from [here](https://github.com/atmh/ip/releases/tag/v0.1).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 📜 

And it is **FREE!**

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

atmh committed:
Level-1: Greet, Echo, Exit

atmh committed:
Added documentation

atmh committed:
Minor changes to documentation

atmh committed:
Level-2: Add, List

atmh committed:
Level-3: Mark as Done

atmh committed:
Level-4: ToDos, Events, Deadlines

atmh committed:
A-TextUiTesting: Automated Text UI Testing

atmh committed:
Level-5: Handle Errors

atmh committed:
Level-6: Delete

atmh committed:
changed if-else to switch

atmh committed:
renamed task subclasses

atmh committed:
renamed task subclasses

atmh committed:
Migrated TaskList class to Echo class

atmh committed:
update printing of response

atmh committed:
fix code style

atmh committed:
Improve coding standard

atmh committed:
Add save and read file

atmh committed:
Add date and time

atmh committed:
Merge branch 'branch-Level-7'

atmh committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Echo.java

atmh committed:
Fix save format

atmh committed:
Add more OOP

atmh committed:
Add packages

atmh committed:
Add junit test

atmh committed:
Add jar file

atmh committed:
Add javadoc

atmh committed:
Improve coding standard

atmh committed:
Add find command

atmh committed:
Merge branch 'branch-A-JavaDoc'

atmh committed:
Merge branch 'branch-A-CodingStandard'

atmh committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/echo/parser/Parser.java
#	src/main/java/echo/ui/Ui.java

atmh committed:
Add java doc

atmh committed:
Update showHelp() with find command

atmh committed:
Remove wildcard imports

atmh committed:
Clean up manifest

atmh committed:
Add A-Gradle

atmh committed:
Merge branch 'branch-A-Gradle'

atmh committed:
Add GUI

atmh committed:
Merge branch 'branch-Level-10'

1001mei commented on src/main/java/echo/parser/Parser.java:
```diff
@@ -0,0 +1,256 @@
+            switch (command) {
+            case ListCommand.COMMAND:
+                return prepareListCommand();
+            case TodoCommand.COMMAND:
+            case DeadlineCommand.COMMAND:
```
Perhaps you could add explicit `//Fallthrough` comment for these `case` statements as stated in Java coding standard?

atmh replied:
Thanks for pointing that out, will do!

1001mei commented on src/main/java/echo/parser/Parser.java:
```diff
@@ -0,0 +1,256 @@
+                                + "\n" + "E.g. 2019-10-15 1800");
+                    }
+                }
+            case MarkCommand.COMMAND:
+            case UnmarkCommand.COMMAND:
```
Perhaps you could add explicit `//Fallthrough` comment here also.

1001mei commented on src/main/java/echo/parser/Parser.java:
```diff
@@ -0,0 +1,256 @@
+        return isExit;
+    }
+
+    /**
+     * Prepare list command.
+     *
```
According to Java coding standard, the first sentence in method header comments should start in the form `Returns ...`, `Adds ...` (not `Return` or `Returnning` etc.).
Maybe you can consider changing it to better adhere to the standard? I noticed this same issue in several other places too.
```suggestion
    /**
     * Prepares list command.
     *
```

atmh replied:
Good eye! I will change them as you suggested 👍.

1001mei commented on src/main/java/echo/task/Task.java:
```diff
@@ -0,0 +1,64 @@
+        this.isDone = true;
+    }
+
+    /**
+     * Marks task.
+     */
```
I believe this should be Unmarks task?

atmh replied:
oops 😅

1001mei commented on src/main/java/echo/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,58 @@
+    private final String DESCRIPTION;
+
+    /** Date and time of deadline command. */
+    private final LocalDateTime DATE_TIME;
```
I like how your names for classes and attributes are very clear. 👍

1001mei commented on src/main/java/echo/parser/Parser.java:
```diff
@@ -0,0 +1,256 @@
+        return new HelpCommand();
+    }
+
+    /**
+     * Prepare bye command.
+     *
+     * @return ByeCommand.
+     */
```
I like how you wrote java doc for all classes and methods, and even for private methods! Good effort! 💯

1001mei reviewed (commented):
Overall, I found your code very clean and easy to read, just that there are a few minor coding standard violations here and there. Keep up the good work!

atmh committed:
Fix minor coding standard violations

atmh committed:
Rename test functions

atmh committed:
Update mainClassName to echo.Launcher

There can be classpath issues if echo.main.Echo is used.

Launcher class is a workaround for classpath issues.

Let's update mainClassName to echo.Launcher.

atmh committed:
Add assert to Parser localDateTime

atmh committed:
Merge pull request #1 from atmh/branch-A-Assertions

Add assert to Parser localDateTime

atmh committed:
Remove obvious unnecessary comments

The code is self-explanatory.

Obvious unnecessary comments are removed.

Let's not repeat the description in a comment for
self-explanatory code.

atmh committed:
Improve code quality

atmh committed:
Merge pull request #2 from atmh/branch-A-CodeQuality

Improve code quality

atmh committed:
Set up CI

atmh committed:
Fix gradlew permission

atmh committed:
Fix checkstyle

atmh committed:
Add detect duplicates

atmh committed:
Clean up extra whitespace

atmh committed:
Remove unused import

atmh committed:
Merge pull request #3 from atmh/branch-BCD-Extension

C-DetectDuplicates

atmh committed:
Fix load error message

cashewnade commented on src/main/java/echo/Echo.java:
```diff
@@ -0,0 +1,77 @@
+    private final Ui UI;
+
+    /**
+     * Constructor for Echo.
```
Perhaps "Constructs the Echo object" would be more clear? (For constructors)

cashewnade commented on src/main/java/echo/command/DeleteCommand.java:
```diff
@@ -0,0 +1,55 @@
+     *
+     * @throws EchoException If input is invalid.
+     */
+    @Override
```
If the method has an @/Override tag (ie the method is overriding a method from the parent class), I believe you can leave out the documentation :)

cashewnade commented on src/main/java/echo/parser/Parser.java:
```diff
@@ -0,0 +1,280 @@
+            switch (command) {
+            case ListCommand.COMMAND:
+                return prepareListCommand();
+            case TodoCommand.COMMAND:
```
Indentation levels for the switch commands are correct! Consider adding a line between each case to enhance readability! :)

cashewnade commented on src/main/java/echo/parser/Parser.java:
```diff
@@ -0,0 +1,280 @@
+                    throw new EchoException("Please specify the task number. Eg. mark 1");
+                }
+
+                // Parse second input into an int, throws NumberFormatException if it's not an int.
```
The in-text comments have been really helpful in understanding your code. Just a small note to take, "Parses" might be more appropriate! :)

cashewnade commented on src/main/java/echo/parser/Parser.java:
```diff
@@ -0,0 +1,280 @@
+    }
+
+    /**
+     * Is user exited.
```
"Returns if user is exited" would be better!

cashewnade commented on src/main/java/echo/parser/Parser.java:
```diff
@@ -0,0 +1,280 @@
+     *
+     * @return True if user is exited; False otherwise.
+     */
+    public static boolean isExit() {
```
Would making isExit() return instanceof ByeCommand be more appropriate?

atmh replied:
hmm, that could be better, i'll look into incorporating it in.

cashewnade commented on src/main/java/echo/storage/Storage.java:
```diff
@@ -0,0 +1,139 @@
+                String type = splitVerticalBar[0];
+                String desc = splitVerticalBar[2];
+
+                // Add task to tasks based on type.
```
"Adds" would be more appropriate!

atmh replied:
Thanks for identifying the grammatical errors

cashewnade reviewed (commented):
LGTM! Your code was very neat and easy to read, informative JavaDocs written for almost all the methods, regardless of private or public. No major violations of the coding standards. Hope to see the full product! :)

atmh committed:
Fix comments

atmh committed:
Improve Code Quality

atmh committed:
Improve GUI

atmh committed:
Merge pull request #5 from atmh/branch-A-BetterGui

Improve GUI

atmh committed:
Add user guide

atmh committed:
Merge pull request #6 from atmh/branch-A-UserGuide

Add user guide

atmh committed:
Update README.md

atmh committed:
Crop Ui.png

atmh committed:
Fix save file path
